File: C# Programs/MergeSort.cs
using System;

//A Merge Sort implementation
//Time Complexity: O(nlog(n))

public class MergeSort
{
    public static void Main(string[] args)
    {
        int[] array = { 5, 9, 2, 3, 6, -7 };
        Console.WriteLine("Unsorted: [{0}]", string.Join(", ", array));

        var sortedArr = MergeSort(array, 0, array.Length - 1);

        Console.WriteLine("Sorted (Merge Sort): [{0}]", string.Join(", ", sortedArr));
    }

    private static int[] MergeSort(int[] array, int start, int end)
    {
        if (start < end)
        {
            int middle = (end + start) / 2;
            int[] leftArr = MergeSort(array, start, middle);
            int[] rightArr = MergeSort(array, middle + 1, end);
            int[] mergedArr = MergeArray(leftArr, rightArr);
            return mergedArr;
        }
        return new int[] { array[start] };
    }

    private static int[] MergeArray(int[] leftArr, int[] rightArr)
    {
        int[] mergedArr = new int[leftArr.Length + rightArr.Length];

        int leftIndex = 0;
        int rightIndex = 0;
        int mergedIndex = 0;

        // Traverse both arrays simultaneously and store the smallest element of both to mergedArr
        while (leftIndex < leftArr.Length && rightIndex < rightArr.Length)
        {
            if (leftArr[leftIndex] < rightArr[rightIndex])
            {
                mergedArr[mergedIndex++] = leftArr[leftIndex++];
            }
            else
            {
                mergedArr[mergedIndex++] = rightArr[rightIndex++];
            }
        }

        // If any elements remain in the left array, append them to mergedArr
        while (leftIndex < leftArr.Length)
        {
            mergedArr[mergedIndex++] = leftArr[leftIndex++];
        }

        // If any elements remain in the right array, append them to mergedArr
        while (rightIndex < rightArr.Length)
        {
            mergedArr[mergedIndex++] = rightArr[rightIndex++];
        }

        return mergedArr;
    }
}

File: Java Programs/merge_sort.java
package rec3;

public class merge_sort {
	
	public static void mergeSort(int[] input){
		
        sort(input , 0, input.length-1);
		
	}
    
      
    public static void merge(int input[], int beg, int mid, int end){  
  
        int l = mid - beg + 1;  
        int r = end - mid;  
  
        int LA[] = new int [l];  
        int RA[] = new int [r];  
  
        for (int i=0; i<l; i++)  
             LA[i] = input[beg + i];  
  
        for (int j=0; j<r; j++)  
             RA[j] = input[mid + 1+ j];  
        
        int i = 0, j = 0;  
        int k = beg;  
         
        while (i<l&&j<r)  
        {  
            if (LA[i] <= RA[j])  
            {  
                input[k] =LA[i];  
                i++;  
            }  
            else  
            {  
                input[k] = RA[j];  
                j++;  
            }  
            k++;  
        }  
        
        while (i<l)  
        {
            input[k] = LA[i];  
            i++;  
            k++;  
        }  
  
        while (j<r)  
        {  
            input[k] = RA[j];  
            j++;  
            k++;  
        }  
    }  
  
    static void sort(int input[], int p, int r)  
    {  
        if (p<r)  
        {  
            int q = (p+r)/2;  
            sort(input, p,q);  
            sort(input , q+1, r);  
            merge(input, p, q, r); 
        }  
    }  

	
	public static void main(String[] args)
	{
		int input[]= {2,1,5,2,3};
		sort(input,0,input.length-1);
		for(int i=0;i<input.length;i++)
		{
			System.out.print(input[i]+" ");
		}
	}

}
